DDD-1.md

```diff
@@ -0,0 +1,173 @@
+
+Currently, the "real" name of the source database is detected right after connecting to the database at the connection level. In order to make the connection reusable across multiple databases (https://github.com/sugarcrm/debezium/issues/10), we need to:
+
+1. Detect the "real" name of all databases the task is working with.
+2. Make the "real" name available to all components that use it right now.
+
+The most natural place for that seems to be the `SqlServerTaskPartition` and its `Provider` being introduced as part of #15.
+
+### Use fully-qualified names for all tables and procedures in Debezium connector for SQL Server ([#10](https://github.com/sugarcrm/debezium/issues/10))
+
+See details in [#10](https://github.com/sugarcrm/debezium/issues/10). This is where the connector-level partitions defined in the beginning come into play. Given a partition, the connector will be able to call `partition.getDatabase()` and perform a query / execute a statement on a connection that isn't scoped to any specific database.
+
+### Introduce `Partitioned*` APIs ([#13](https://github.com/sugarcrm/debezium/pull/13), [#16](https://github.com/sugarcrm/debezium/issues/16), [#18](https://github.com/sugarcrm/debezium/issues/18))
+
+####  `Partitioned*` interfaces
+
+Despite the fact that there are multiple connectors that can benefit from being aware, some connectors are nonpartitioned by their nature. For instance, the MySQL connector consumes all the changes from a nonpartitioned binlog, and it would make sense to have its API and the implementation affected by the changes being proposed as little as possible.
+
+In order to accomodate partition-aware and -unaware components under the same framework, it's proposed to introduce the `Partitioned` analogues of the following interfaces:
+
+1. `SnapshotChangeEventSource` → `PartitionedSnapshotChangeEventSource`
+2. `StreamingChangeEventSource` → ``PartitionedStreamingChangeEventSource`
+3. `ChangeEventSourceFactory` → `PartitionedChangeEventSourceFactory`
+
+Each of their methods that implements capturing data will have two more arguments added:
+
+1. `<P extends TaskPartition> taskPartition` to identify which task partition the method call should use.
+2. `<O extends OffsetContext> offsetContext` the offset context that corresponds to the partition.
+
+This way, each of the connectors could implement the relevant of the two APIs. It will also allow connectors to transition from one API to another and allow the user to chose the mode via configuration.
+
+#### Affect on framework-level classes
+
+Some framework-level classes that are reused by multiple connectors and depend on the interfaces above will have to be reworked to satisfy both of the APIs. It could be either one class per interface or one class implementing both interfaces. The latter looks more reasonable since their implementatin will be the same for the most part.
```

Comment:
```suggestion
Some framework-level classes that are reused by multiple connectors and depend on the interfaces above will have to be reworked to satisfy both of the APIs. It could be either one class per interface or one class implementing both interfaces. The latter looks more reasonable since their implementation will be the same for the most part.
```